cg/apps/sequencing_metrics_parser/parsers/bcl_convert.py

```diff
@@ -0,0 +1,204 @@
+"""This module parses metrics for files generated by the BCLConvert tool using the dragen hardware."""
+import csv
+import xml.etree.ElementTree as ET
+import logging
+from pathlib import Path
+from typing import List, Union
+from cg.io.controller import ReadFile
+from cg.constants.demultiplexing import SampleSheetHeaderColumnNames
+from cg.constants.constants import FileFormat
+from cg.apps.sequencing_metrics_parser.models.bcl_convert import (
+    BclConvertAdapterMetrics,
+    BclConvertDemuxMetrics,
+    BclConvertQualityMetrics,
+    BclConvertRunInfo,
+    BclConvertSampleSheet,
+)
+
+
+LOG = logging.getLogger(__name__)
+
+
+class BclConvertMetricsParser:
+    def __init__(
+        self,
+        bcl_convert_quality_metrics_path: Path,
+        bcl_convert_demux_metrics_file_path: Path,
+        bcl_convert_adapter_metrics_file_path: Path,
+        bcl_convert_sample_sheet_file_path: Path,
+        bcl_convert_run_info_file_path: Path,
+    ) -> None:
+        """Initialize the class."""
+        self.quality_metrics_path: Path = bcl_convert_quality_metrics_path
+        self.demux_metrics_path: Path = bcl_convert_demux_metrics_file_path
+        self.adapter_metrics_path: Path = bcl_convert_adapter_metrics_file_path
+        self.sample_sheet_path: Path = bcl_convert_sample_sheet_file_path
+        self.run_info_path: Path = bcl_convert_run_info_file_path
+        self.quality_metrics: List[BclConvertQualityMetrics] = self.parse_quality_metrics_file()
+        self.demux_metrics: List[BclConvertDemuxMetrics] = self.parse_demux_metrics_file()
+        self.adapter_metrics: List[BclConvertAdapterMetrics] = self.parse_adapter_metrics_file()
+        self.sample_sheet: List[BclConvertSampleSheet] = self.parse_sample_sheet_file()
+        self.run_info: BclConvertRunInfo = self.parse_run_info_file()
+
+    def parse_quality_metrics_file(
+        self,
+    ) -> List[BclConvertQualityMetrics]:
+        """Parse the BCL convert metrics file with read pair format into a BclConvertQualityMetrics model."""
+        LOG.info(f"Parsing BCLConvert metrics file: {self.quality_metrics_path}")
+        parsed_metrics: List[BclConvertQualityMetrics] = []
+        metrics_reader = ReadFile.get_content_from_file(
+            file_format=FileFormat.CSV, file_path=self.quality_metrics_path, read_to_dict=True
+        )
+        for row in metrics_reader:
+            parsed_metrics.append(
+                BclConvertQualityMetrics(
+                    lane=int(row["Lane"]),
```

Comment:
I think you can move the type conversion to a common per type function in the model and set when initializing. The the model can handle the conversion and we do not have to bother the caller to something about the file types nor the model requirements.

Similar to this, but using set in the name and tailored to the conversion you want to do:

Borrowed from Atlas doc parsing
```
    @validator("revision_date", "yearly_revision_date", always=True)
    def validate_revision_date(cls, value: str) -> datetime.date:
        if isinstance(value, str):
            return datetime.datetime.strptime(value, "%Y-%m-%d").date()
        return value
```

Reply:
Implemented, made eveyrhting much clearer!